### PR TITLE
Persistent Descriptor Pool

### PR DESCRIPTION
**Pull request by TeBoring**
Previously, because it's possible that different php requests may have conflicting proto message (different message with the same fully qualified name), it is required that the internal descriptor pool is refreshed after each request, which hurts performance.

This change implements persistent descriptor pool. Users needs to make sure there is no conflicting messages by themselves.

By default, this feature is disabled. To enable it, add ```protobuf.keep_descriptor_pool_after_request=1``` into php.ini. Or on command line, add ```-d protobuf.keep_descriptor_pool_after_request=1```.

### CODE REVIEW

File: php/ext/google/protobuf/protobuf.c [hunk 1]
```diff
@@ -241,30 +331,21 @@ static void php_proto_hashtable_descriptor_release(zval* value) {
...
-  zend_hash_init(ce_to_php_obj_map, 16, NULL, HASHTABLE_VALUE_DTOR, 0);
+static void test_release(void* value) {
+  void* ptr = value;
+}
+#endif
 
-  ALLOC_HASHTABLE(proto_to_php_obj_map);
-  zend_hash_init(proto_to_php_obj_map, 16, NULL, HASHTABLE_VALUE_DTOR, 0);
+static initialize_persistent_descriptor_pool(TSRMLS_D) {
+  upb_inttable_init(&upb_def_to_desc_map_persistent, UPB_CTYPE_PTR);
```
Comment (stanley-cheung) [inline php/ext/google/protobuf/protobuf.c]: What does `upb_inttable_init` do? This seems central to the theme about creating the persistent pool.

File: php/ext/google/protobuf/protobuf.c [hunk 2]
```diff
@@ -276,23 +357,79 @@ static PHP_RINIT_FUNCTION(protobuf) {
...
+
+  ALLOC_HASHTABLE(ce_to_php_obj_map);
+  zend_hash_init(ce_to_php_obj_map, 16, NULL, HASHTABLE_VALUE_DTOR, 0);
+
+  generated_pool = NULL;
+  generated_pool_php = NULL;
+  internal_generated_pool_php = NULL;
+
+  if (!PROTOBUF_G(keep_descriptor_pool_after_request)) {
+    initialize_persistent_descriptor_pool(TSRMLS_C);
```
Comment (stanley-cheung) [inline php/ext/google/protobuf/protobuf.c]: So just to confirm - the logic here is that: if the new php.ini config entry is not set (old behavior), we do this `initialize_persistent_descriptor_pool` (which I suppose is just current behavior), for every PHP request? (hence this `RINIT` function.

File: php/ext/google/protobuf/protobuf.c (hunk 1 above)
Comment (TeBoring) [reply php/ext/google/protobuf/protobuf.c]: It initializes a custom hashtable implemented by upb.

File: php/ext/google/protobuf/protobuf.c [hunk 3]
```diff
@@ -113,6 +112,40 @@ PHP_PROTO_HASHTABLE_VALUE get_def_obj(const void* def) {
...
 }
 
+void add_msgdef_desc(const upb_msgdef* m, DescriptorInternal* desc) {
+  upb_inttable_insertptr(&upb_def_to_desc_map_persistent,
+                         m, upb_value_ptr(desc));
+}
+
+DescriptorInternal* get_msgdef_desc(const upb_msgdef* m) {
+  upb_value v;
+#ifndef NDEBUG
```
[the remark this answers is not in the document]
Comment (TeBoring) [reply php/ext/google/protobuf/protobuf.c]: NDEBUG is a standard macro during compiling, which means no debug mode.
This is somehow exposed by upb. We don't actually depend on it. Just to let upb happy.

File: php/ext/google/protobuf/protobuf.c (hunk 2 above)
Comment (TeBoring) [reply php/ext/google/protobuf/protobuf.c]: Yes

**Pull request merged**